JSON serializer for task graphs

jl-wynen opened the pull request:
First step of #92

Please start with the new notebook to read about the reasoning behind the format and see examples.

SimonHeybrock commented on docs/user-guide/index.md:
```diff
@@ -8,4 +8,5 @@ maxdepth: 2
 getting-started
 parameter-tables
 generic-providers
+serialization
```
Does not match page title

SimonHeybrock commented on src/sciline/__init__.py:
```diff
@@ -10,7 +10,7 @@
 
 from . import scheduler
-from ._provider import UnboundTypeVar
+from ._provider import ArgSpec, Provider, ProviderKind, ProviderLocation, UnboundTypeVar
```
Why are these top-level now?

jl-wynen replied:
Leftover from a helper function that I removed again.

SimonHeybrock reviewed (commented):
Before a complete review, some general questions/concerns:

- Why are data nodes serialized? I thought we just want to serialize the task graph (which may eventually include serializing the *input* data, or a hash thereof).
- It is not clear why we would want special handling for param tables. Is the only reason a potential desire to reconstruct the same `TaskGroup` from the JSON? Otherwise I would feel that the special handling just overcomplicates things. I do not really see a need for handling this any different from "regular" providers.

jl-wynen commented:
> Why are data nodes serialized? I thought we just want to serialize the task graph (which may eventually include serializing the input data, or a hash thereof).

Why does `visualize` show them? Without the data nodes, it becomes harder to understand what a provider returns and how that relates to arguments of other providers. We would basically have to write the signature of providers into the nodes.

> It is not clear why we would want special handling for param tables. Is the only reason a potential desire to reconstruct the same TaskGroup from the JSON? Otherwise I would feel that the special handling just overcomplicates things. I do not really see a need for handling this any different from "regular" providers.

What is a 'TaskGroup'?

table providers are handled the same as others. It is the data nodes that have special handling because they encode which table entry they represent. If this wasn't the case, we would have a lot of duplicate, e.g., `float(int)` nodes in the graph.

SimonHeybrock commented:
> > Why are data nodes serialized? I thought we just want to serialize the task graph (which may eventually include serializing the input data, or a hash thereof).
> 
> Why does `visualize` show them? Without the data nodes, it becomes harder to understand what a provider returns and how that relates to arguments of other providers. We would basically have to write the signature of providers into the nodes.

`visualize` is not about provenance. I am not saying serializing the result types is a bad idea, but then I thought it should be part of the `'compute'` nodes, not represented as `'data'` nodes. But if we would distinguish between `'input'` nodes (which may get serialized values eventually) and `'data'` nodes (values never serialized) that would be equivalent.


> > It is not clear why we would want special handling for param tables. Is the only reason a potential desire to reconstruct the same TaskGroup from the JSON? Otherwise I would feel that the special handling just overcomplicates things. I do not really see a need for handling this any different from "regular" providers.
> 
> What is a 'TaskGroup'?

`TaskGraph`

> table providers are handled the same as others. It is the data nodes that have special handling because they encode which table entry they represent. If this wasn't the case, we would have a lot of duplicate, e.g., `float(int)` nodes in the graph.

Not an issue if my first item is taken into account.

jl-wynen commented:
> visualize is not about provenance.

This doesn't answer my question. Why is it useful to show the types as separate nodes in visualize but not for provenance tracking?

> But if we would distinguish between 'input' nodes (which may get serialized values eventually) and 'data' nodes (values never serialized) that would be equivalent.

What would be equivalent?

> > table providers are handled the same as others. It is the data nodes that have special handling because they encode which table entry they represent. If this wasn't the case, we would have a lot of duplicate, e.g., float(int) nodes in the graph.

> Not an issue if my first item is taken into account.

Well, the special handling would be moved from data to provider nodes. But it would still be there.

SimonHeybrock commented:
> > visualize is not about provenance.
> 
> This doesn't answer my question. Why is it useful to show the types as separate nodes in visualize but not for provenance tracking?

Not sure why you don't see having the types in the visualization is useful. But computation wise, the graph of providers provides a full definition.

> > But if we would distinguish between 'input' nodes (which may get serialized values eventually) and 'data' nodes (values never serialized) that would be equivalent.
> 
> What would be equivalent?

Adding an extra attribute in a compute node saying "I am computing type X" is equivalent to adding a data node between this compute node and all dependent compute nodes.

> > > table providers are handled the same as others. It is the data nodes that have special handling because they encode which table entry they represent. If this wasn't the case, we would have a lot of duplicate, e.g., float(int) nodes in the graph.
> 
> > Not an issue if my first item is taken into account.
> 
> Well, the special handling would be moved from data to provider nodes. But it would still be there.

Why, how? I don't see the need for special handling. You'd just have the same compute node N times (all with a unique `id`).

jl-wynen commented:
> Why, how? I don't see the need for special handling. You'd just have the same compute node N times (all with a unique id).

And identical return types. Or not easily computer-readable return types (Without special handling, the serialized type becomes, e.g., `"builtins.float(builtins.int:0)"` and that still has special handling in `key_full_qualname`)

SimonHeybrock commented:
> > Why, how? I don't see the need for special handling. You'd just have the same compute node N times (all with a unique id).
> 
> And identical return types. Or not easily computer-readable return types (Without special handling, the serialized type becomes, e.g., `"builtins.float(builtins.int:0)"` and that still has special handling in `key_full_qualname`)

What is the problem with identical return types? You will get N data nodes (all with a unique id).

jl-wynen commented:
> > > Why, how? I don't see the need for special handling. You'd just have the same compute node N times (all with a unique id).
> > 
> > 
> > And identical return types. Or not easily computer-readable return types (Without special handling, the serialized type becomes, e.g., `"builtins.float(builtins.int:0)"` and that still has special handling in `key_full_qualname`)
> 
> What is the problem with identical return types? You will get N data nodes (all with a unique id).

Not when you merge data and provider nodes. And even if not, you get a bunch of nodes that are identical up to the id. So you know they represent different things, but cannot tell what they represent.

SimonHeybrock commented:
> > > > Why, how? I don't see the need for special handling. You'd just have the same compute node N times (all with a unique id).
> > > 
> > > 
> > > And identical return types. Or not easily computer-readable return types (Without special handling, the serialized type becomes, e.g., `"builtins.float(builtins.int:0)"` and that still has special handling in `key_full_qualname`)
> > 
> > 
> > What is the problem with identical return types? You will get N data nodes (all with a unique id).
> 
> Not when you merge data and provider nodes.

I think that discussion is orthogonal and does not change anything.

> And even if not, you get a bunch of nodes that are identical up to the id. So you know they represent different things, but cannot tell what they represent.

So? I thought this is about provenance. If you look at an arbitrary non-param-table graph you also have no idea what it represents, without knowing which input data was used, do you?

jl-wynen commented:
> So? I thought this is about provenance. If you look at an arbitrary non-param-table graph you also have no idea what it represents, without knowing which input data was used, do you?

But the graph has all graph-related info. Yes, parameters are missing, this was part of the requirements for this task. But with tables, even the graph is incompletely specified:
```mermaid
graph LR
A1("A(Label)") --> B1("B(Label)") --> C
A2("A(Label)") --> B2("B(Label)") --> C
A1("A(Label)") --> D
```
Which part of the param table is used by D and which by C?

To be honest, I don't fully understand your objection. Which part of the information in teh graph should be omitted?

SimonHeybrock commented:
> But the graph has all graph-related info. Yes, parameters are missing, this was part of the requirements for this task. But with tables, even the graph is incompletely specified:
> [...]
> Which part of the param table is used by D and which by C?

The value "1" from the table is used by D, both "1" and "2" are used by C? "1" and "2" are "input" nodes.

```mermaid
graph LR
1-->A1
2-->A2
A1("A") --> B1("B") --> C
A2("A") --> B2("B") --> C
A1("A") --> D
```

nvaytet commented on docs/user-guide/provenance.ipynb:
```diff
@@ -0,0 +1,279 @@
+    "It can, however,  only capture part of the actual pipeline.\n",
+    "For example, it only shows the structure of the graph and contains the names of functions and types.\n",
+    "But it does not encode the implementation of those functions or types.\n",
+    "Thus, the graph can only be correctly reconstructed in an environment that contains the same software that was used to write the graph.\n",
```
Would be nice to show an example of loading the json, building a new task graph from it and computing the result?
But I guess you've only implemented the saving of graphs, not loading them?

jl-wynen replied:
I considered writing a loader. But that needs more info than is in the graph right now (parameters, package versions). And it's non-trivial because it needs some importlib finagling, so I wouldn't necessarily show it in the docs.

nvaytet commented on src/sciline/visualize.py:
```diff
@@ -83,18 +83,16 @@ def to_graphviz(
+        if kind == 'series':
             # Example: Series[RowId, Material[Country]] -> RowId, Material[Country]
             return name.partition('[')[-1].rpartition(']')[0]
-        return sgname
```
This used to return the first name after splitting with `[`, but is now returning the whole thing?
Is that intentional?

jl-wynen replied:
No, it's an oversight. It means that before all subgraphs that return `Series[*]` were merged. Is this intentional?

nvaytet commented on tests/serialize/json_test.py:
```diff
@@ -0,0 +1,254 @@
+    tg = TaskGraph(graph=graph, keys=str)
+    res = tg.serialize()
+    schema = json_schema()
+    jsonschema.validate(res, schema)
```
Along the lines of my first comment, I think it would be good if we could have something a bit like a round-trip test where we save the graph, then load it again and compute something with it and check that it's the same as what the original graph gives?

But this implies we have something that can load saved graphs...

jl-wynen commented:
> > But the graph has all graph-related info. Yes, parameters are missing, this was part of the requirements for this task. But with tables, even the graph is incompletely specified:
> > [...]
> > Which part of the param table is used by D and which by C?
> 
> The value "1" from the table is used by D, both "1" and "2" are used by C? "1" and "2" are "input" nodes.
> 
> ```mermaid
> graph LR
> 1-->A1
> 2-->A2
> A1("A") --> B1("B") --> C
> A2("A") --> B2("B") --> C
> A1("A") --> D
> ```

Those input nodes correspond to the `p_table_cell` nodes in the PR and are not regular parameter nodes. So would you preserve those but remove the index information from all dependent nodes? This would mean that you have to walk the graph to find that information.

SimonHeybrock commented:
> > > But the graph has all graph-related info. Yes, parameters are missing, this was part of the requirements for this task. But with tables, even the graph is incompletely specified:
> > > [...]
> > > Which part of the param table is used by D and which by C?
> > 
> > 
> > The value "1" from the table is used by D, both "1" and "2" are used by C? "1" and "2" are "input" nodes.
> > 
> >   ```mermaid
> > graph LR
> > 1-->A1
> > 2-->A2
> > A1("A") --> B1("B") --> C
> > A2("A") --> B2("B") --> C
> > A1("A") --> D
> > ```
> >   
> >     
> >   
> >   
> > 
> >   
> > 
> >       graph LR
> > 1-->A1
> > 2-->A2
> > A1("A") --> B1("B") --> C
> > A2("A") --> B2("B") --> C
> > A1("A") --> D
> 
> Those input nodes correspond to the `p_table_cell` nodes in the PR and are not regular parameter nodes. So would you preserve those but remove the index information from all dependent nodes? This would mean that you have to walk the graph to find that information.

Yes, absolutely! Remember that Sciline's parameter tables are just a *convenient way of building task graphs*. They do however have little to do with the resulting task graph and actual computation. As this task is about provenance and reproducibility (independent of Sciline), we have to avoid exposing Sciline "implementation details" in the JSON representation.

In fact I would be tempted to argue that a graph that was made without parameter tables "by hand" should result in the same JSON (modulo something related to the provider that makes the series).

jl-wynen commented:
Implemented a new schema with these changes:
- no support for param tables
- merged data and provider nodes
- renamed type -> out
- function nodes now encode the order of arguments

SimonHeybrock commented on src/sciline/serialize/_json.py:
```diff
@@ -0,0 +1,146 @@
+        edges.append(edge)
+        kwargs[name] = edge['id']
+
+    node = {
+        'id': node_id,
+        'kind': 'function',
+        'label': provider_name(provider),
+        'function': provider_full_qualname(provider),
+        'out': key_full_qualname(key),
+        'args': args,
+        'kwargs': kwargs,
+    }
```
Without having looked at everything, I think this looks much better now (with the Sciline-specifics removed) 👍 

Question: With the `out` as part of the node, we make all dependent nodes linked to the `id` of the provider. Maybe that is not so great, and we should have an `id` for the data, i.e., re-introduce "data" nodes so they can have their own `id`? Or do you think it does not matter?

jl-wynen replied:
I had a long discussion with Max about this. Separating data and functions is in principle cleaner. It would even remove the need for the parameter kind. And if we ever need to attach attributes to data, we could do so without affecting function nodes.

However, it makes the graph harder to read and the whole process a little more complicated. So the thought was to go with what I implemented now and see how that goes and change later if need be. We still have a little time to figure things out.

SimonHeybrock replied:
> Separating data and functions is in principle cleaner.

I thought it would be cleaner especially since we also have input nodes already, a special kind of data node.

> However, it makes the graph harder to read

By whom?

> and the whole process a little more complicated.

Wouldn't it simply mean that providers need to add two nodes, the compute node and the output data node?

jl-wynen replied:
> I thought it would be cleaner especially since we also have input nodes already, a special kind of data node.

This is what I meant by 

> > It would even remove the need for the parameter kind

> > However, it makes the graph harder to read
>
> By whom?
>
> > and the whole process a little more complicated.
>
> Wouldn't it simply mean that providers need to add two nodes, the compute node and the output data node?

By humans because they need to follow two edges to find how data relates to each other instead of one. And by computers because they need to take into account the two fundamentally different node types. E.g., for reconstructing a task graph, the reader must extract pairs of nodes for each task instead of just one.

For writing, this is not that much more complicated and well contained.

jl-wynen replied:
> But they are connected through a direct edge? Is it actually complicated enough to warrant combining compute and data nodes?

No. This makes a miniscule difference on all points unless we foresee adding more fields to data nodes.
I'm happy to change back. I mainly went this way because you objected to having separate notes initially.

SimonHeybrock replied:
> I mainly went this way because you objected to having separate notes initially.

I did not object, I mainly asked about it! And my main concern there was the discrepancy between inputs (then stored as functions) and data nodes.

It seems Aiida has data nodes? https://aiida.readthedocs.io/projects/aiida-core/en/latest/howto/visualising_graphs.html

jl-wynen replied:
Yes, they do. This links into what I hinted at with storing extra metadata in data nodes. If we, e.g., stored intermediate results in a database or in files, we could link to them here, kind of like Aiida does.

SimonHeybrock reviewed (approved):
I think this looks good now.
